addons/token-exchange/token_exchanger_agent.go

```diff
@@ -0,0 +1,168 @@
+package addons
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/openshift/library-go/pkg/controller/controllercmd"
+	"github.com/openshift/library-go/pkg/controller/factory"
+	"github.com/openshift/library-go/pkg/operator/events"
+	"github.com/openshift/library-go/pkg/operator/resource/resourceapply"
+	"github.com/spf13/cobra"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/version"
+	"k8s.io/client-go/informers"
+	corev1informers "k8s.io/client-go/informers/core/v1"
+	"k8s.io/client-go/kubernetes"
+	corev1lister "k8s.io/client-go/listers/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/clientcmd"
+	"k8s.io/klog/v2"
+	"open-cluster-management.io/addon-framework/pkg/lease"
+)
+
+const (
+	TokenExchangeName = "tokenexchange"
+)
+
+func NewAgentCommand() *cobra.Command {
+	o := NewAgentOptions()
+	cmd := controllercmd.
+		NewControllerCommandConfig(TokenExchangeName, version.Info{Major: "0", Minor: "1"}, o.RunAgent).
+		NewCommand()
+	cmd.Use = TokenExchangeName
+	cmd.Short = "Start the token exchange addon agent"
+
+	o.AddFlags(cmd)
+	return cmd
+}
+
+// AgentOptions defines the flags for agent
+type AgentOptions struct {
+	HubKubeconfigFile string
+	SpokeClusterName  string
+}
+
+// NewAgentOptions returns the flags with default value set
+func NewAgentOptions() *AgentOptions {
+	return &AgentOptions{}
+}
+
+func (o *AgentOptions) AddFlags(cmd *cobra.Command) {
+	flags := cmd.Flags()
+	flags.StringVar(&o.HubKubeconfigFile, "hub-kubeconfig", o.HubKubeconfigFile, "Location of kubeconfig file to connect to hub cluster.")
+	flags.StringVar(&o.SpokeClusterName, "cluster-name", o.SpokeClusterName, "Name of spoke cluster.")
+}
+
+// RunAgent starts the controllers on agent to process work from hub.
+func (o *AgentOptions) RunAgent(ctx context.Context, controllerContext *controllercmd.ControllerContext) error {
```

Comment:
How do you run the agent and which Context do you use? How do you handle termination signals?

Reply:
All done for us by `openshift/library-go`.
It runs the agent with right context and handles termination as well.